Creating C and Python bindings for all functionality used by Repl.  Refactoring Repl to call through Python by default

luketpeterson opened the pull request:
Creating C and Python bindings for all functionality used by Repl.
Refactoring Repl to call through Python by default.

Changes include:
* Adding Environment object to Rust, C & Python for non-interactive configuration management (module search paths, config files, etc.)
* Adding SyntaxNode parsing API to C & Python
* Adding RunnerState / incremental MeTTa runner API to C & Python
* Adding parse error accessor to C & Python, instead of panicking on invalid syntax
* Refactoring REPL app, so that it calls MeTTa through Python by default.  NOTE: I am preserving the direct Rust path for the "no_python" path because I plan to unify the "python" and "no_python" paths in the future, once we can link HyperonPy in a static build.  (when we address https://github.com/trueagi-io/hyperon-experimental/issues/283)
* Changing REPL default to Python, and requiring explicit "no_python" feature to build REPL without Python dependency
* Removing ctor to initialize logger, because it makes sense to init the logger from the environment initialization

luketpeterson commented:
This PR should address:
https://github.com/trueagi-io/hyperon-experimental/issues/439
https://github.com/trueagi-io/hyperon-experimental/issues/386
https://github.com/trueagi-io/hyperon-experimental/issues/301

And perhaps others.

vsbogd commented on c/src/metta.rs:
```diff
@@ -569,3 +939,169 @@ pub extern "C" fn metta_load_module(metta: *mut metta_t, name: *const c_char) {
+// way it is for now, but create a nicer API for Python using kwargs
+
+/// cbindgen:ignore
+static CURRENT_ENV_BUILDER: Mutex<(EnvInitState, Option<EnvBuilder>)> = std::sync::Mutex::new((EnvInitState::Uninitialized, None));
```
May be implement builder in C in a straightforward way by passing instance of the builder as an argument to the setters? Is there any other reason to use global `CURRENT_ENV_BUILDER` except removing one of the arguments?

luketpeterson replied:
To me it was a matter of API ergonomics & surface area.  Fundamentally configuring the platform environment is a stateful operation (ie. It's done only once at startup), so having stateful API functions doesn't actually sacrifice any flexibility.

Global data feels evil in an object-oriented language.  But C isn't an object oriented language, and sometimes making global state feel like an object is counter-productive.  This is just my opinion, however.

If you feel strongly, I can wrap this in something that feels more object-oriented.  I just wanted to keep the API surface area to a minimum and blend in with the conventions of the host language (C in this case).

vsbogd replied:
> To me it was a matter of API ergonomics & surface area. Fundamentally configuring the platform environment is a stateful operation (ie. It's done only once at startup), so having stateful API functions doesn't actually sacrifice any flexibility.

From my perspective it is a question of whether we need two different environments in the same process. For example if we use global instance of the `Environment` the unit tests which use different environment state cannot run in parallel. It is not fully the same but still relevant example is introducing reentrant versions of the C API functions into POSIX to overstep single thread implementation restrictions.

luketpeterson replied:
That’s a very good point.  I didn’t consider the use case of multiple C / Python threads racing to init the Environment at the same time.  I figured the client app would perform the init before spawning child threads, but I see that’s not always possible.

I reworked the C API (and Python by extension, although it's hidden in Python) to use a builder object here: 586e2b2

As to whether we want to allow multiple environments to be active in the same process at the same time, that's a slightly different question.  Do you think that's important?

luketpeterson replied:
In dc15fc3, I extensively reworked the environment API to permit custom environments to be initialized and used from any host language, and I provided conveniences to initialize a test environment that will not load / nor create any platform configuration files.

The singleton environment is now purely a convenience.

vsbogd reviewed (commented):
Luke, thanks for the such amount of work. Didn't look at the REPL code changes yet, but posting current comments to not slow this down.

luketpeterson commented on lib/src/metta/environment.rs:
```diff
@@ -0,0 +1,193 @@
+
+        let mut env = self.env;
+
+        if !self.no_cfg_dir {
```
The reason I ended up with this design was that the logic is actually tri-state, and I wanted the API to be very explicit.

The possible states are:
1. User will accept the OS-supplied default config directory
2. User wants to specify a custom config directory
3. User wants to tell MeTTa not to use any config directory whatsoever, and to not touch the file system at all.

I felt that third option was qualitatively different from the other two, and deserved a separate flag.

Another option would be an Enum with an embedded Path, which is what I would have gone for in pure Rust, but that has a lot of surface area because there is no similarly-behaved language construct in C nor in Python.

luketpeterson commented on lib/src/metta/environment.rs:
Ah ok.  I can move it.  To me it was always a little fuzzy.

luketpeterson replied:
Moved in 3be64f5

luketpeterson commented on lib/src/metta/environment.rs:
```diff
@@ -0,0 +1,193 @@
+
+const DEFAULT_INIT_METTA: &[u8] = include_bytes!("init.default.metta");
+
+static PLATFORM_ENV: std::sync::OnceLock<Environment> = std::sync::OnceLock::new();
```
Right.  The idea is to make it low-friction for the caller.  So the caller doesn't need to keep track of the environment as an object.

A caller can initialize the platform environment explicitly as the first thing they do.  Otherwise, the default environment will be automatically initialized lazily the first time it's needed.

As for unit tests, we could init the environment with the `no_cfg_dir` option, which would prevent the any `init.metta` file from running.  But doing that at the top of every test is a little tedious.  One option is to make a test preamble that we always run at the top of every test.  Another option is to make a test-specific `Metta::new_rust` function.

I think I'd prefer the test preamble approach because it lets us handle other places where the test environment differs from normal execution.

vsbogd replied:
> The idea is to make it low-friction for the caller. So the caller doesn't need to keep track of the environment as an object.
> A caller can initialize the platform environment explicitly as the first thing they do. Otherwise, the default environment will be automatically initialized lazily the first time it's needed.

The straightforward (not necessary right) way to do this is to make an instance of the `Environment` a part of the `MeTTa` instance. Which disadvantages do you see in this approach?

> But doing that at the top of every test is a little tedious. One option is to make a test preamble that we always run at the top of every test. Another option is to make a test-specific Metta::new_rust function.

Yes, I agree doing this in each test is too verbose. First thing which comes into my mind is to make a test-specific `Metta::new_rust()` function. I am not sure what is preamble approach. Could you please share a link to some example of this?

luketpeterson replied:
It's a non-issue now, after the refactor in https://github.com/trueagi-io/hyperon-experimental/commit/dc15fc359ebb7b7dad9014891da563a6dfda709d the MeTTa runners can be initialized with a test environment.

luketpeterson commented on lib/src/metta/runner/mod.rs:
```diff
@@ -118,6 +154,12 @@ impl Metta {
                 runner.run(&mut SExprParser::new(program.as_str()))
                     .map_err(|err| format!("Cannot import module, path: {}, error: {}", path.display(), err))?;
+
+                // TODO: This is a hack. We need a way to register tokens at module-load-time, for any module
+                if path.to_str().unwrap() == "stdlib" {
+                    register_rust_tokens(self);
+                }
```
My opinion is that we should do what's required to make `register_rust_tokens` be compatible with any runner that loads the rust stdlib.  Is there a fundamental reason this is impossible?

vsbogd replied:
Yes, I agree in general that we are going to reuse Rust stdlib implementation for other host languages. The question was more about whether we need to do this step right now. But it is a minor point if it works.

luketpeterson commented on lib/src/lib.rs:
```diff
@@ -7,10 +7,3 @@ pub mod space;
-
-use ctor::ctor;
-
-#[ctor]
-fn on_load() {
-    common::init_logger(false);
-}
```
Good catch.  My thinking was that the first call to init the Environment (including by creating a MeTTa runner) will init the logger.  But that wasn't happening in the code.  So I fixed it in b80a23f.

At some point I think we can add custom logger integration, as per https://github.com/trueagi-io/hyperon-experimental/issues/146  But I don't know how high the priority is on that.

luketpeterson commented on python/hyperon/runner.py:
All of the setters are implemented in C & Python.  The getters are not.

The reason for this is that the Rust API has two very different clients: 1.) Outside callers like the REPL, etc. and 2.) The rest of the Rust library itself (for example to access parameters when initializing a Runner)

The function of the Environment object is to allow outside callers to set parameters that affect the way MeTTa is configured.  So communication is mainly in one direction: from the outside caller into MeTTa.

I'm not conceptually against adding additional getters where they make sense.  But I don't feel like most of the information in the environment is useful to an outside caller.  But if you disagree we can add it.

vsbogd replied:
Main reason why I am asking this is a question should Python and Rust API be kept as similar as possible or not. Usually I (as a library user) expect that API of the library in different languages contains the same set of abstractions. It is not always convenient and possible from API development perspective. There is an inevitable difference between Rust and Python paradigms. In particular this is the reason why Python `Atom` uses inheritance instead of sum of the types which is used in Rust version. But I think we need to try keep APIs similar as much as possible.

After your comment my understanding that you introduced methods in `RunnerState` because in Python you can easily do it while in Rust borrow checker makes such API inconvenient. But I don't see anything wrong with Rust approach here and I think it would be simpler to support this API if Python and Rust version will be similar. At least it is a general approach I am trying to stick to.

This question is general and it is not specific to this PR so may be we need to discuss it with a wider audience. But at least I would like to know your opinion and experience regarding this.

luketpeterson replied:
My feeling is that Rust, because it's the language of the core, will always have a more expansive set of functionality.  This is definitional, because everything exposed outside Rust has to call into a public Rust implementation somewhere in the core.

So I don't think API parity with Rust is a worthwhile goal - we'll never get there for the above reason - but also I think it's not the point of an API.

I think a more practical goal is to expose the functionality that will be useful to clients in the target host language.  And in fact, we want to expose the minimum set of functionality because API surface area is potential maintenance work down the road. 

So I agree with your choices to expose polymorphic atom types in Python, because Python supports better object-oriented function dispatch than straight C.  My Python and C API choices were also attempting to be guided by best practices on each host language.

On the specifics of `RunnerState`, I don't have a strong opinion.  It would be just as easy for me to transform it back to match Rust.  I just felt the alternative formulation was a better fit for the language.

vsbogd replied:
> So I don't think API parity with Rust is a worthwhile goal ...  And in fact, we want to expose the minimum set of functionality because API surface area is potential maintenance work down the road.

I agree that we don't need expose the whole Rust API without a reason. My comment above was about using the same set of abstractions rather than equality.

My opinion about this is based on personal experience of using libraries with bindings for the different languages. I saw both kinds of libraries where API was really different while still implemented the same functionality and similar and usually I felt that latter is more natural. One can easily switch from one language to another without learning API from the scratch. I would put it in words like: "if both implementations feel natural in a language then choose the way which is similar to a core library".

If you feel that the way it is implemented in Python is more natural to Python and we cannot do it in a Rust in the similar way then lets leave it as is.

luketpeterson replied:
In [457d757](https://github.com/trueagi-io/hyperon-experimental/pull/450/commits/457d757ea9cb5b2e46c1feec197b46c8a5072791), I ended up changing the Rust RunnerState API to match what I had in Python.  So now they are harmonized across languages.

Necr0x0Der commented on python/hyperon/base.py:
```diff
@@ -340,6 +380,19 @@ def parse(self, tokenizer):
         catom = self.cparser.parse(tokenizer.ctokenizer)
         return Atom._from_catom(catom) if catom is not None else None
 
+    def parse_err(self):
```
All Python exceptions are wrapped into `Error`-expression in MeTTa. However, this is a way for passing unpredicted exceptions to MeTTa. If some error is typical and expected (as syntax mistakes), then it might be better to have a dedicated error message in MeTTa (maybe, without inclusion of references to the implementation language). OTOH, if there is a chance that `parse` will be called outside the MeTTa runner, then throwing a Python exception can be convenient.  Thus, I don't have a strong preference. And I didn't check this PR in practice, so I'm ok with having it as is and refining this later if necessary.

vsbogd replied:
I believe wrapping the parsing error into a `Error`-expression should be done inside `import` atom implementation (or other possible MeTTa expressions which interpret the MeTTa program as a text). This is a question about Python API and I believe we can choose either way. I think it is a minor thing, just think that exceptions are more natural to Python as a way of error processing.

Necr0x0Der replied:
Yes, this is what I also meant. I agree that Python functions should throw exceptions if they are not direct implementations of grounded functions for MeTTa, and these exceptions can be caught and turned into specific `Error`-expressions in the grounded functions like `import!`. The only thing in this case is that if `parse` throws an exception for the main script, what will happen?

vsbogd replied:
Caller should decide whether to catch exception and handle error or allow exception be re-thrown further. In case of main script it should catch the exception and print the error to the user I believe.

luketpeterson replied:
This is implemented in a3808f1

Necr0x0Der commented on python/tests/scripts/f1_imports.metta:
```diff
@@ -1,13 +1,19 @@
+; contains an atom wrapping the space of stdlib.
 ; The type of this atom is the same as of `&self`
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+; QUESTION: Are space implementations really required to preserve ordering of atoms?
```
We cannot impose the requirement of preserving the atoms order to all spaces. It will definitely be not the case for many spaces like DAS or potentially neural or hypervector spaces. It is convenient to have the order preservation feature for the in-RAM program space, and some people already rely on this (although we warn them that this may change). It's ok to break order preservation (and it might be less confusing to do this asap), but in this case, we need to provide means to do what these people do with `superpose` and `collapse` (e.g., sequential execution of side-effect functions, sequential traversal of MeTTa expressions, etc.).

Necr0x0Der replied:
I looked at this question in the context and would like to add that this question is irrelevant to this situation. The script is parsed and evaluated sequentially. The comment means that there is already `&stdlib` atom placed in `&self` space already at the beginning of the script evaluation. It doesn't mean and doesn't matter that `&stdlib` atom is placed at the beginning of the space. Basically, how this comment is changed in this PR sounds more confusing to me than it was before.  Maybe, the comment requires further clarification and the question can be removed as irrelevant.

Necr0x0Der commented on python/tests/scripts/f1_imports.metta:
```diff
@@ -1,13 +1,19 @@
 ; The type of this atom is the same as of `&self`
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+; QUESTION: Are space implementations really required to preserve ordering of atoms?
+; QUESTION / TODO: Is there a better way to find this sub-space than `get-atoms`?
+;    Now that we load the init.metta file as a module, the previous comment about the
+;    &self space containing only one atom at the beginning is no longer true
```
My idea was to have a library manager (which was partly implemented) interfaced to MeTTa, so one could have an analog of Python `__modules__`, etc. Apparently, we have a huge issue regarding imports / modules, and this question should be considered as a part of this large issue. BTW, I think that we need some way to have systematic design discussions instead of considering different yet related issues separately. Maybe, having tracking issues on github may work, but maybe there is a better technical way to structure these discussions.

luketpeterson replied:
It's a non-issue now because the test environment no longer loads the init.metta.  I hope we can find a better way to test module loading than relying on specific atom ordering in the space, but for now I have reverted the test changes back to the previous state in https://github.com/trueagi-io/hyperon-experimental/commit/dc15fc359ebb7b7dad9014891da563a6dfda709d

vsbogd reviewed (commented):
Luke, I am very sorry for a late review. Thanks for adding ability to initialize runner with instance of the environment :+1: I propose to go further with fixing the rest of comments and merging it.

One major question to me here is should we keep some shared instance of the environment or just replace it by asking user to create a default instance if it is needed?

luketpeterson commented on c/src/metta.rs:
```diff
@@ -335,6 +582,10 @@ pub extern "C" fn get_atom_types(space: *const space_t, atom: *const atom_ref_t,
 // MeTTa Intperpreter Interface
 // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
 
+//QUESTION: It feels like a runner_state_t and a step_result_t are getting at the same functionality,
```
As we discussed, I left the separate interpreter API in for now.

I added the capabilities to the RunnerState such that it could emulate the behavior of the Interpreter API with a temporary MeTTa, but it is much higher overhead to call through a RunnerState than just to call the interpreter directly.

We may revisit this if grounded atom execution is reworked to depend on a context with access to a MeTTa runner.  https://github.com/trueagi-io/hyperon-experimental/issues/410

luketpeterson commented on lib/src/metta/environment.rs:
```diff
@@ -0,0 +1,229 @@
+impl Environment {
+
+    /// Returns a reference to the shared "platform" Environment
+    pub fn platform_env() -> &'static Self {
```
It is a global environment instance, initialized once, and available anywhere with a simple accessor.

The reason I chose the word "platform" is that my ideas was that it was the configuration set up during program initialization or the default if none was explicitly set.  Any of words you suggested (global, shared, common) would be ok, but I don't feel like they quite express the purpose of the object.  I'd say `shared` is probably my favorite among them.

luketpeterson replied:
After some more thought, I went with "common".  Changed in 9029f66

luketpeterson commented on lib/src/metta/environment.rs:
```diff
@@ -0,0 +1,229 @@
+        // A better fix is to fork a "child runner" with access to the same space and tokenizer,
+        // but an updated search path.  This is really hard to implement currently given the ImportOp
+        // actually owns a reference to the runner it's associated with.  However this must be fixed soon.
+        env.extra_include_paths.push(".".into());
```
What you suggest is the ultimate goal.  The problem is that some of the operation atoms in the present design each contain internal references to the runner.  So when we add a new runner with different search paths and/or a different working dir, we'd need to reload all stdlib operation atoms for the child runner, which is definitely not ideal.

When we address https://github.com/trueagi-io/hyperon-experimental/issues/410 then we can remove this "." from the search paths.

luketpeterson commented on lib/src/metta/runner/mod.rs:
```diff
@@ -64,32 +68,79 @@ pub struct RunnerState<'a> {
+    ///
+    /// NOTE: If `env_builder` is `None`, the platform environment will be used
+    /// NOTE: This function does not load any stdlib atoms, nor run the [Environment]'s 'init.metta'
+    pub fn new_with_space(space: DynSpace, tokenizer: Shared<Tokenizer>, env_builder: Option<EnvBuilder>) -> Self {
```
Yes.  I felt like having the built environment in an owned state complicated the API (especially in other languages) without much gain.

So now `Environment` is like `Path`; ie. the API user can have access &Environment references, and call methods on it, but they can't own a naked `Environment` outside the hyperon crate.

luketpeterson commented on python/hyperon/runner.py:
```diff
@@ -61,11 +90,53 @@ def load_py_module(self, name):
+        except:
+            return None
+
+    def load_py_module_from_mod_or_file(self, mod_name):
```
Ah ok.  I felt like it kinda walks the line.  It's there to support `import-py!`, but it does exactly what the caller would expect if it's called normally.

So I don't know if it's better private or not.

luketpeterson replied:
What do you think @vsbogd ?  I can make an argument to leave it public and a separate argument that it should be private.  Do you have an opinion?

vsbogd replied:
Lets keep it as is, I don't have a strong opinion here and it is minor anyway.

luketpeterson commented on lib/src/metta/runner/mod.rs:
```diff
@@ -64,32 +68,79 @@ pub struct RunnerState<'a> {
+    ///
+    /// TODO: When we are able to load the Rust stdlib before the Python stdlib, which requires value-bridging,
+    ///     we can refactor this function to load the appropriate stdlib(s) and simplify the init process
+    pub fn init(&self) {
```
My original thinking was that this `init` method is a placeholder for any and all things that need to be done to finalize the initialization of the runner.  Currently that's just the `init.metta`, but I was thinking that a "finalize" call might be a good hook to have if we add more runner features in the future.

However, I have been thinking about your comments more, and I think the whole initialization process is pretty ugly.  In particular, I hate the fact that `init_rust` is a separate pathway, and that there is so much exposed surface area for implementing a different MeTTa runner initialization process.  So I'm going to try and rework this API a little bit.

luketpeterson replied:
Ok.  In 3ac2582 I added `Metta::new_with_stdlib_loader` to specify a hook to add a custom stdlib.  Now it's not up to the downstream dependencies to initialize the MeTTa runner in the right order.

I also replaced `Metta::new_rust` with just `Metta::new`, which is a lot cleaner, IMO.

luketpeterson commented on python/hyperon/atoms.py:
```diff
@@ -27,6 +27,9 @@ def __repr__(self):
         return hp.atom_to_str(self.catom)
 
+    def is_error(self):
+        return hp.atom_is_error(self.catom)
```
Moved in 205b4d0

luketpeterson commented:
I believe I have addressed all of the feedback in this PR.

The only remaining issue is whether or not `runner.load_py_module_from_path` should be private, and any new issues resulting from the changes I made.

vsbogd commented on python/tests/test_extend.py:
```diff
@@ -18,10 +18,6 @@ def test_extend(self):
              [ValueAtom(5)],
              [ValueAtom('B')]])
-        self.assertEqual(
-              metta.run('! &runner')[0][0].get_object().value,
```
Did you remove it by accident or it doesn't work after changes.

luketpeterson replied:
I removed it intentionally because the runner object is actually a handle to the runner.  Two runners are different boxed objects but they point to the same runner.

luketpeterson replied:
Do you think checking MeTTa pointer equality is important enough that it makes sense to add a "metta_eq" function to C?  My thinking is that it probably doesn't matter except for this test.

luketpeterson replied:
Ultimately I figured comparing the referenced runner is worthwhile functionality.  Added in 8254d3f

luketpeterson commented on lib/src/metta/runner/mod.rs:
```diff
@@ -63,57 +54,69 @@ enum MettaRunnerMode {
+    /// Create and initialize a MeTTa interpreter with a language-specific stdlib
+    ///
+    /// NOTE: pass `None` for `env_builder` to use the common environment
+    pub fn new_with_stdlib_loader<F>(loader: F, env_builder: Option<EnvBuilder>) -> Metta
```
As we discussed, a custom space argument is needed because the runner may wish to use a different top-level space implementation for &self.  I added back the space parameter in 790ad1e

However, what is the use case for creating a MeTTa with an existing Tokenizer?

vsbogd replied:
You are right that after import is added injecting some custom tokenizer doesn't make much sense. Formally it is a part of a parser and  I don't like current way of decoupling this. I believe we can remove a way of injecting tokenizer via constructor safely.

luketpeterson commented on lib/src/metta/mod.rs:
```diff
@@ -42,6 +41,56 @@ pub const UNIFY_SYMBOL : Atom = sym!("unify");
+        None => EMPTY_SYMBOL,
+    };
+    if let Some(err_code) = err_code {
+        Atom::expr([ERROR_SYMBOL, err_atom, err_code, Atom::sym(message)])
```
The idea was that sometimes both an `err_code` and a custom message are desirable.  The caller can branch on an err code in a way that they can't reliably branch on a human-readable message.  But a human-readable message can specify additional information, including composing a message with specific details about the error.

luketpeterson commented on c/src/metta.rs:
```diff
@@ -205,53 +207,32 @@ pub extern "C" fn sexpr_parser_new(text: *const c_char) -> sexpr_parser_t {
-            parser.err_string = err_cstring.into_raw();
-            atom_t::null()
-        }
+        Err(err) => error_atom(None, None, err).into()
```
After our discussion on the call, I've re-added the error accessor back to the C API, and call it from Python.  d5253d7

vsbogd reviewed (approved):
Thanks Luke, I approved it, please feel free to merge when you are ready.

vsbogd commented:
Thanks Luke, I merged it

Necr0x0Der commented:
Thanks Luke. Looks cool and working!
